source/server/listener_manager_impl.cc

```diff
@@ -88,19 +90,47 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, ListenerManag
       local_drain_manager_(parent.factory_.createDrainManager()) {
   // TODO(htuch): Support multiple filter chains #1280, add constraint to ensure we have at least on
   // filter chain #1308.
-  ASSERT(config.filter_chains().size() == 1);
-  const auto& filter_chain = config.filter_chains()[0];
-
-  listener_scope_ =
-      parent_.server_.stats().createScope(fmt::format("listener.{}.", address_->asString()));
+  ASSERT(config.filter_chains().size() >= 1);
+
+  // Skip lookup and update of the SSL Context if there is only one filter chain
+  // and it doesn't enforce any SNI restrictions.
+  bool skip_context_update;
+  if (config.filter_chains().size() == 1 &&
+      config.filter_chains()[0].filter_chain_match().sni_domains().empty()) {
+    skip_context_update = true;
+  } else {
+    skip_context_update = false;
+  }
 
-  if (filter_chain.has_tls_context()) {
-    Ssl::ServerContextConfigImpl context_config(filter_chain.tls_context());
-    ssl_context_ = parent_.server_.sslContextManager().createSslServerContext(*listener_scope_,
-                                                                              context_config);
+  size_t has_tls = 0;
+  size_t has_stk = 0;
+  for (const auto& filter_chain : config.filter_chains()) {
+    std::vector<std::string> sni_domains(filter_chain.filter_chain_match().sni_domains().begin(),
+                                         filter_chain.filter_chain_match().sni_domains().end());
+    if (filter_factories_.empty() || !filter_chain.filters().empty()) {
+      // Only one filter chain can have filters until multiple filter chains are properly supported.
+      RELEASE_ASSERT(filter_factories_.empty());
+      filter_factories_ = parent_.factory_.createFilterFactoryList(filter_chain.filters(), *this);
+    }
+    if (filter_chain.has_tls_context()) {
+      Ssl::ServerContextConfigImpl context_config(filter_chain.tls_context());
+      tls_contexts_.emplace_back(parent_.server_.sslContextManager().createSslServerContext(
+          name_, sni_domains, *listener_scope_, context_config, skip_context_update));
+      has_tls++;
+      if (filter_chain.tls_context().has_session_ticket_keys()) {
+        has_stk++;
+      }
+    }
   }
 
-  filter_factories_ = parent_.factory_.createFilterFactoryList(filter_chain.filters(), *this);
+  // TODO(PiotrSikora): allow filter chains with mixed use of Session Ticket Keys.
+  // This doesn't work right now, because BoringSSL uses "session context" (initial SSL_CTX that
+  // accepted connection, before SNI update) for session related stuff, including Session Ticket
+  // callback, which is going to be called iff it's set on the initial SSL_CTX, even if it's not
+  // set on the current SSL_CTX that doesn't have any Session Ticket Keys configured.
+  ASSERT(has_stk == 0 || has_stk == has_tls);
```

Comment:
Shouldn't we actually throw an EnvoyException here so that the config load fails?

Reply:
I was thinking the same... but we're guarding those invalid cases with `ASSERT()` in most of the places with incomplete API v2 support, e.g. at beginning of this function.

Reply:
Hmm. I feel like that was put there during transition in which we were only converting JSON and protected by schema. Now that we are allowing people to send us pure v2, I think we need better checks here. Should we convert this to exceptions that would fail config load? @htuch ?

Reply:
Yes, I think we should have exceptions in all places where the proto constraints won't be providing a guard.

Reply:
Done.